### PR TITLE
Add `PIP_COMPATIBILITY.md` to document known deviations from `pip`

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Closes https://github.com/astral-sh/uv/issues/2023.

### CODE REVIEW

**Comment by charliermarsh** [comment]
Still draft, missing one or two things, and need to adjust the README.

File: PIP_COMPATIBILITY.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+[#12470](https://github.com/pypa/pip/issues/12470), etc.). uv's pre-release handling is
+_intentionally_ limited and _intentionally_ requires user opt-in for pre-releases, to ensure
+correctness.
+
+In the future, uv _may_ support pre-release identifiers in transitive dependencies. However, it's
+likely contingent on evolution in the Python packaging specifications. The existing PEPs [do not
+cover "dependency resolution"](https://discuss.python.org/t/handling-of-pre-releases-when-backtracking/40505/17)
+and are instead focused on behavior for a _single_ version specifier. Further, the [original intent](https://discuss.python.org/t/handling-of-pre-releases-when-backtracking/40505/20)
+of the spec was _not_ to support cases in which pre-releases are _required_ for resolution despite
+a lack of pre-release specifiers. As such, there are unresolved questions around the correct and
```
Comment (konstin) [inline PIP_COMPATIBILITY.md:131]: I'd skip the original intent part, back there dependency resolution wasn't a thing in python so PEP 440 doesn't account for it, and instead move the link to the pip behavior paragraph above.

**Comment by charliermarsh** [comment]
Okay, ready for review. The goal here is for this to something we can link to when folks have questions about sub-items in the document.

File: PIP_COMPATIBILITY.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+private `requests` would conflict with the public [`requests`](https://pypi.org/project/requests/)
+on PyPI.
+
+When uv searches for a package across multiple indexes, it will iterate over the indexes in order
+(preferring the `--extra-index-url` over the default index), and stop searching as soon as it
+finds a match. This means that if a package exists on multiple indexes, uv will limit its
+candidate versions to those present in the first index that contains the package.
+
+(`pip`, meanwhile, will combine the candidate versions from all indexes, and select the best
+version from the combined set.)
```
Comment (konstin) [inline PIP_COMPATIBILITY.md:161]: I'd also add that pip makes no guarantees about the order (e.g. https://github.com/pypa/pip/issues/5045#issuecomment-369521345)

File: PIP_COMPATIBILITY.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+## Virtual environments by default
+
+`uv pip install` and `uv pip sync` are designed to work with virtual environments by default.
+
+Specifically, uv will always install packages into the currently active virtual environment, or
+search for a virtual environment named `.venv` in the current directory or any parent directory
+(even if it is not activated).
+
+This differs from `pip`, which will install packages into a global environment if no virtual
+environment is active, and will not search for inactive virtual environments.
```
Comment (konstin) [inline PIP_COMPATIBILITY.md:183]: I'd add a word of warning here about the user and global env here, no isolation between different projects and different cli tools and the global env on linux can break system packages.

File: PIP_COMPATIBILITY.md [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+uv tends to be stricter than `pip`, and will often reject packages that `pip` would install.
+For example, uv omits packages with invalid version specifiers in its metadata, which `pip` plans
+to do in a [future release](https://github.com/pypa/pip/issues/12063).
+
+In some cases, uv implements lenient behavior for popular packages that are known to have
+specific spec compliance issues.
+
+If uv rejects a package that `pip` would install due to a spec violation, the best course of
+action is to first attempt to install a newer version of the package; and, if that fails, to report
+the issue to the package maintainer.
```
Comment (konstin) [inline PIP_COMPATIBILITY.md:210]: :100:

File: PIP_COMPATIBILITY.md [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+- [`--trusted-host`](https://github.com/astral-sh/uv/issues/1339)
+- [`--user`](https://github.com/astral-sh/uv/issues/2077)
+- [`--no-build-isolation`](https://github.com/astral-sh/uv/issues/1715)
+
+If you encounter a missing option or subcommand, please search the issue tracker to see if it has
+already been reported, and if not, consider opening a new issue. Feel free to upvote any existing
+issues to convey your interest.
+
+## Resolution strategy
```
Comment (konstin) [inline PIP_COMPATIBILITY.md:228]: I'd move this higher, this is a common confusion for newcomers

File: PIP_COMPATIBILITY.md [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    # via fastapi
+typing-extensions==4.10.0
+    # via
+    #   fastapi
+    #   pydantic
+    #   pydantic-core
+```
+
+When uv resolutions differ from `pip` in undesirable ways, it's often a sign that the specifiers
+are too loose, and that the user should consider tightening them. For example, in the case of
```
Comment (konstin) [inline PIP_COMPATIBILITY.md:308]: I'd make this more concrete: Please add `>=` bounds to your packages.

**Review by konstin — APPROVED** [review]
:scroll: :scroll: :scroll:

**Review by zanieb — APPROVED** [review]
Nice!

File: PIP_COMPATIBILITY.md (hunk 6 above)
Comment (zanieb) [reply PIP_COMPATIBILITY.md:308]: Isn't this clear in the example?

File: PIP_COMPATIBILITY.md [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+
+If dependency resolution fails due to a transitive pre-release, uv will prompt the user to
+re-run with `--prerelease=allow`, to allow pre-releases for all dependencies.
+
+Alternatively, you can add the transitive dependency to your `requirements.in` file with
+pre-release specifier (e.g., `flask>=2.0.0rc1`) to opt in to pre-release support for that specific
+dependency.
+
+In sum, uv needs to know upfront whether the resolver should accept pre-releases for a given
+package. `pip`, meanwhile, will generally respect pre-release identifiers in transitive
```
Comment (notatallshaw) [inline PIP_COMPATIBILITY.md:116]: FYI, I expanded on this a bit here: https://github.com/astral-sh/uv/issues/1641#issuecomment-1981402429

It's not important for uv, but certainly anything stronger than "generally" would be wrong, could even be "sometimes".

**Review by BurntSushi — APPROVED** [review]

File: PIP_COMPATIBILITY.md (hunk 7 above)
Comment (charliermarsh) [reply PIP_COMPATIBILITY.md]: Thanks!

File: PIP_COMPATIBILITY.md [hunk 8]
```diff
@@ -0,0 +1,322 @@
+# Compatibility with `pip` and `pip-tools`
+
+uv is designed as a drop-in replacement for common `pip` and `pip-tools` workflows.
+
+Informally, the intent is such that existing `pip` and `pip-tools` users can switch to uv without
+making meaningful changes to their packaging workflows; and, in most cases, swapping out
+`pip install` for `uv install` should "just work".
```
[the remark this answers is not in the document]
Comment (charliermarsh) [reply PIP_COMPATIBILITY.md:7]: Thanks! https://github.com/astral-sh/uv/pull/2262

**Pull request merged**